File: Assets/Tests/EditMode/PlayerShotEventCodecTests.cs
using System;
using NUnit.Framework;
using TopDownRoguelike.Networking.Protocol;
using UnityEngine;

namespace TopDownRoguelike.Tests.EditMode
{
    public sealed class PlayerShotEventCodecTests
    {
        [Test]
        public void EncodeDecode_PreservesAllFields()
        {
            var source =
                new PlayerShotEvent(
                    7u,
                    42u,
                    1.25f,
                    -2.5f,
                    0.6f,
                    0.8f);

            byte[] encoded =
                PlayerShotEventCodec.Encode(source);

            Assert.That(
                encoded.Length,
                Is.EqualTo(PlayerShotEventCodec.PayloadSize));

            PlayerShotEvent decoded =
                PlayerShotEventCodec.Decode(encoded);

            Assert.That(decoded.PlayerId, Is.EqualTo(7u));
            Assert.That(decoded.ShotSequence, Is.EqualTo(42u));
            Assert.That(decoded.OriginX, Is.EqualTo(1.25f));
            Assert.That(decoded.OriginY, Is.EqualTo(-2.5f));
            Assert.That(decoded.DirectionX, Is.EqualTo(0.6f));
            Assert.That(decoded.DirectionY, Is.EqualTo(0.8f));
        }

        [Test]
        public void Decode_RejectsTruncatedPayload()
        {
            Assert.Throws<ArgumentException>(
                () =>
                    PlayerShotEventCodec.Decode(
                        new byte[PlayerShotEventCodec.PayloadSize - 1]));
        }

        [Test]
        public void Decode_RejectsTrailingBytes()
        {
            Assert.Throws<ArgumentException>(
                () =>
                    PlayerShotEventCodec.Decode(
                        new byte[PlayerShotEventCodec.PayloadSize + 1]));
        }

        [Test]
        public void Decode_RejectsNonFiniteOrigin()
        {
            var source =
                new PlayerShotEvent(
                    7u,
                    42u,
                    0.0f,
                    0.0f,
                    1.0f,
                    0.0f);

            byte[] encoded =
                PlayerShotEventCodec.Encode(source);

            PacketCodec.WriteNetworkUInt32(
                encoded,
                8,
                0x7FC00000u);

            Assert.Throws<ArgumentException>(
                () =>
                    PlayerShotEventCodec.Decode(encoded));
        }

        [Test]
        public void Decode_RejectsNonFiniteDirection()
        {
            var source =
                new PlayerShotEvent(
                    7u,
                    42u,
                    0.0f,
                    0.0f,
                    1.0f,
                    0.0f);

            byte[] encoded =
                PlayerShotEventCodec.Encode(source);

            PacketCodec.WriteNetworkUInt32(
                encoded,
                16,
                0x7F800000u);

            Assert.Throws<ArgumentException>(
                () =>
                    PlayerShotEventCodec.Decode(encoded));
        }
    }
}

File: NetworkServer/tests/PlayerShotEventCodecTests.cpp
#include "protocol/PlayerShotEventCodec.h"

#include <cmath>
#include <cstdint>
#include <exception>
#include <iostream>
#include <stdexcept>
#include <vector>

namespace
{
    void Require(
        const bool condition,
        const char* message
    )
    {
        if (!condition)
        {
            throw std::runtime_error(message);
        }
    }

    void RequireNear(
        const float actual,
        const float expected,
        const float epsilon,
        const char* message
    )
    {
        if (std::fabs(actual - expected) > epsilon)
        {
            throw std::runtime_error(message);
        }
    }
}

void EncodeDecodePreservesAllFields()
{
    tdr::protocol::PlayerShotEvent source{};
    source.playerId = 7U;
    source.shotSequence = 42U;
    source.originX = 1.25F;
    source.originY = -2.5F;
    source.directionX = 0.6F;
    source.directionY = 0.8F;

    const std::vector<std::uint8_t> encoded =
        tdr::protocol::PlayerShotEventCodec::Encode(
            source
        );

    Require(
        encoded.size() ==
        tdr::protocol::kPlayerShotEventPayloadSize,
        "Encoded shot event size did not match."
    );

    const tdr::protocol::PlayerShotEvent decoded =
        tdr::protocol::PlayerShotEventCodec::Decode(
            encoded.data(),
            encoded.size()
        );

    Require(
        decoded.playerId == 7U,
        "Decoded player ID did not match."
    );

    Require(
        decoded.shotSequence == 42U,
        "Decoded shot sequence did not match."
    );

    RequireNear(
        decoded.originX,
        1.25F,
        0.0001F,
        "Decoded origin X did not match."
    );

    RequireNear(
        decoded.originY,
        -2.5F,
        0.0001F,
        "Decoded origin Y did not match."
    );

    RequireNear(
        decoded.directionX,
        0.6F,
        0.0001F,
        "Decoded direction X did not match."
    );

    RequireNear(
        decoded.directionY,
        0.8F,
        0.0001F,
        "Decoded direction Y did not match."
    );
}

void DecodeRejectsInvalidSize()
{
    const std::vector<std::uint8_t> tooShort(
        tdr::protocol::kPlayerShotEventPayloadSize - 1U,
        0U
    );

    bool rejected = false;

    try
    {
        tdr::protocol::PlayerShotEventCodec::Decode(
            tooShort.data(),
            tooShort.size()
        );
    }
    catch (const std::invalid_argument&)
    {
        rejected = true;
    }

    Require(
        rejected,
        "Truncated shot event payload was accepted."
    );
}

int main()
{
    try
    {
        EncodeDecodePreservesAllFields();
        DecodeRejectsInvalidSize();

        std::cout
            << "PlayerShotEventCodecTests passed."
            << std::endl;

        return 0;
    }
    catch (const std::exception& exception)
    {
        std::cerr
            << "[FAIL] "
            << exception.what()
            << std::endl;

        return 1;
    }
}